cdap-app-fabric/src/main/java/io/cdap/cdap/internal/entity/EntityResult.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.entity;
+
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Class for holding paginated search result entities
+ * @param <T> The type of entities
+ */
+public class EntityResult<T> {
+
+  private final Set<T> entities;
+  private final String cursor;
+  private final int offset;
+  private final int limit;
+  private final int totalResults;
+
+  public EntityResult(Set<T> entities, @Nullable String cursor, int offset, int limit,
+                      int totalResults) {
+    this.entities = entities;
```

Comment:
Do a defensive copy of the incoming collection: `this.entities = new LinkedHashSet<>(entities);`

Reply:
If I change this to Collection<? extends T> , a Collections.unmodifiableCollection(entities) should serve the purpose for defensive copying?

Reply:
No. the unmodifiable is just a wrapper, it doesn't do a copy.

Reply:
It should be `Collections.unmodifiableSet(new LinkedHashSet<>(entities));`

Reply:
Done